### PR TITLE
Replace OS-specific stream redirection with pure-Lean `Stream` redirection

### PR DESCRIPTION
**Pull request by Kha**
This avoids the temporary files workaround on macOS and Windows, and makes sure
we can process a `#eval` command and write messages to stdout at the same time.

### CODE REVIEW

File: src/Init/Data/ByteArray/Basic.lean [hunk 1]
```diff
@@ -45,6 +45,22 @@ def set! : ByteArray → (@& Nat) → UInt8 → ByteArray
 def isEmpty (s : ByteArray) : Bool :=
 s.size == 0
 
+/--
+  Copy the slice at `[srcOff, srcOff + len)` in `src` to `[destOff, destOff + len)` in `dest`, growing `dest` if necessary.
+  If `exact` is `false`, the capacity will be doubled when grown. -/
+@[extern "lean_byte_array_copy_slice"]
+def copySlice (src : @& ByteArray) (srcOff : Nat) (dest : ByteArray) (destOff len : Nat) (exact : Bool := true) : ByteArray :=
```
Comment (Kha) [inline src/Init/Data/ByteArray/Basic.lean:52]: @leodemoura This seemed like a highly-desirable primitive

File: src/Init/Data/String/Extra.lean [hunk 2]
```diff
@@ -14,4 +16,13 @@ if s.isNat then
 else
   panic! "Nat expected"
 
+/--
+  Convert a UTF-8 encoded `ByteArray` string to `String`.
+  The result is unspecified if `a` is not properly UTF-8 encoded. -/
```
Comment (Kha) [inline src/Init/Data/String/Extra.lean:21]: @leodemoura I'm assuming that, unlike in Rust, it is currently not `unsafe` to store invalid UTF-8 in a `String`. We might want "checked" alternatives in the future, but for now this should be sufficient.

**Comment by Kha** [comment]
@leodemoura When we add running `Task (IO _)`s (which we should do soon), should we make them inherit the std streams? The output of a task would be lost if the `#eval` that spawned it has already finished, but at least it would not disturb the server stdout if std streams are inherited.

**Comment by leodemoura** [comment]
> @leodemoura When we add running `IO` `Task`s (which we should do soon), should we make them inherit the std streams? The output of a task would be lost if the `#eval` that spawned it has already finished, but at least it would not disturb the server stdout if std streams are inherited.

Yes, this seems to be the way to go.

**Pull request merged**